File: spark-streaming-project/src/main/scala/com/chudichen/spark/dao/CourseClickCountDao.scala
package com.chudichen.spark.dao

import com.chudichen.spark.domain.CourseClickCount
import com.chudichen.spark.utils.HBaseUtil

import scala.collection.mutable.ListBuffer

/**
 * 访问数据库DAO层
 * 今天到现在为止的实战课程的访问量
 *
 * @author chudichen
 * @date 2020-09-01
 */
object CourseClickCountDao {

  val table = "course_click_count"
  val cf = "info"
  val columnQualifier = "click_count"

  /**
   * 像数据库中保存数据
   *
   * @param courseList 课程
   */
  def save(courseList: ListBuffer[CourseClickCount]): Unit = {
    courseList.foreach(course => {
      val rowKey = course.day_course
      val value = course.clickCount
      HBaseUtil.INSTANCE.putData(table, rowKey, cf, columnQualifier, value)
    })
  }

  /**
   * 与旧值进行累加
   *
   * @param courseList 课程列表
   */
  def increase(courseList: ListBuffer[CourseClickCount]): Unit = {
    courseList.foreach(course => {
      val rowKey = course.day_course
      val value = course.clickCount
      val oldValue = HBaseUtil.INSTANCE.getDataLong(table, rowKey, cf, columnQualifier)
      val newValue = oldValue + value
      HBaseUtil.INSTANCE.putData(table, rowKey, cf, columnQualifier, newValue)
    })
  }

  /**
   * 从数据库
   *
   * @param day_course 课程
   * @return
   */
  def get(day_course: String): Long = {
    HBaseUtil.INSTANCE.getDataLong(table, day_course, cf, columnQualifier)
  }
}

File: spring-boot-web/src/main/java/com/chudichen/spark/web/dao/CourseClickCountDao.java
package com.chudichen.spark.web.dao;

import com.chudichen.spark.web.domain.CourseClickCount;
import com.chudichen.spark.web.util.HBaseUtil;
import org.apache.commons.compress.utils.Lists;
import org.apache.hadoop.hbase.client.Result;
import org.apache.hadoop.hbase.client.ResultScanner;
import org.apache.hadoop.hbase.util.Bytes;
import org.springframework.stereotype.Repository;

import java.util.Collections;
import java.util.List;

/**
 * 实战课程访问了Dao
 *
 * @author chudichen
 * @date 2020-09-07
 */
@Repository
public class CourseClickCountDao {

    private static final String CF = "info";
    private static final String COLUMN = "click_count";
    private static final String TABLE = "course_click_count";


    public List<CourseClickCount> query(String dayCourse) {
        ResultScanner scan = HBaseUtil.INSTANCE.scan(TABLE, dayCourse);
        if (scan == null) {
            return Collections.emptyList();
        }

        List<CourseClickCount> courseClickCounts = Lists.newArrayList();
        for (Result result : scan) {
            courseClickCounts.add(new CourseClickCount(Bytes.toString(result.getRow()),
                    Bytes.toLong(result.getValue(Bytes.toBytes(CF), Bytes.toBytes(COLUMN)))));

        }
        return courseClickCounts;
    }
}
